Crashlytics/UnitTests/FIRCLSOnDemandModelTests.m

```diff
@@ -0,0 +1,206 @@
+
+  NSString *name = @"exception_model_report";
+  NSString *reportPath = [self.fileManager.rootPath stringByAppendingPathComponent:name];
+  [self.fileManager createDirectoryAtPath:reportPath];
+
+  FIRCLSInternalReport *report =
+      [[FIRCLSInternalReport alloc] initWithPath:reportPath
+                             executionIdentifier:@"TEST_EXECUTION_IDENTIFIER"];
+  FIRCLSContextInitialize(report, self.mockSettings, self.fileManager);
+}
+
+- (void)tearDown {
+  self.onDemandModel = nil;
+  [[NSFileManager defaultManager] removeItemAtPath:self.fileManager.rootPath error:nil];
+  [super tearDown];
+}
+
+- (void)testIncrementsQueueWhenEventRecorded {
+  FIRExceptionModel *exceptionModel = [self getTestExceptionModel];
+  BOOL success = [self.onDemandModel recordOnDemandExceptionIfQuota:exceptionModel
+                                          withDataCollectionEnabled:YES
+                                         usingExistingReportManager:self.existingReportManager];
+  [self.onDemandModel setQueueToEmpty];
+  // Should record but not submit a report.
+  XCTAssertTrue(success);
+  XCTAssertEqual([self.onDemandModel getOrIncrementOnDemandEventCountForCurrentRun:NO], 1);
+  XCTAssertEqual(self.onDemandModel.getQueuedOperationsCount, 1);
+}
+
+- (void)testCompliesWithDataCollectionOff {
+  FIRExceptionModel *exceptionModel = [self getTestExceptionModel];
+  BOOL success = [self.onDemandModel recordOnDemandExceptionIfQuota:exceptionModel
+                                          withDataCollectionEnabled:NO
+                                         usingExistingReportManager:self.existingReportManager];
+
+  // Should record but not submit a report.
+  XCTAssertTrue(success);
+  // We still count this as an occurred event if it was recorded.
+  XCTAssertEqual([self.onDemandModel getOrIncrementOnDemandEventCountForCurrentRun:NO], 1);
+  XCTAssertEqual([self contentsOfActivePath].count, 2);
+  XCTAssertEqual(self.onDemandModel.getQueuedOperationsCount, 1);
+  XCTAssertEqual([self.onDemandModel.storedActiveReportPaths count], 1);
+}
+
+- (void)testQuotaWithDataCollectionOff {
+  // This test uses the onDemandModel of managerData rather than the mockOnDemandModel to test
+  // integration with existingReportManager
+  FIRExceptionModel *exceptionModel = [self getTestExceptionModel];
+
+  for (int i = 0; i < 10; i++) {
+    BOOL success =
+        [self.managerData.onDemandModel recordOnDemandExceptionIfQuota:exceptionModel
+                                             withDataCollectionEnabled:NO
+                                            usingExistingReportManager:self.existingReportManager];
+
+    XCTAssertTrue(success);
+  }
+
+  // Once we've finished processing, there should be only FIRCLSMaxUnsentReports recorded with the
+  // rest considered dropped. The recorded events should be stored in storedActiveReportPaths which
+  // is kept in sync with the contents of the active path.
+  sleep(60);  // wait for the queue to empty
```

Comment:
We unfortunately can't sleep in tests - I believe there's a way around this though

Reply:
Can't in what way - shouldn't or actually can't?